dbms/src/Dictionaries/PolygonDictionary.cpp

```diff
@@ -0,0 +1,766 @@
+        const std::string & attribute_name,
+        const Columns & key_columns,
+        const DataTypes &,
+        const String & def,
+        ColumnString * const out) const
+{
+    const auto ind = getAttributeIndex(attribute_name);
+    checkAttributeType(name, attribute_name, dict_struct.attributes[ind].underlying_type, AttributeUnderlyingType::utString);
+
+    getItemsImpl<String, StringRef>(
+            ind,
+            key_columns,
+            [&](const size_t, const StringRef value) { out->insertData(value.data, value.size); },
+            [&](const size_t) { return StringRef{def}; });
+}
+
+template <typename AttributeType, typename OutputType, typename ValueSetter, typename DefaultGetter>
+void IPolygonDictionary::getItemsImpl(
+        size_t attribute_ind, const Columns & key_columns, ValueSetter && set_value, DefaultGetter && get_default) const
+{
+    const auto points = extractPoints(key_columns);
+
+    using ColVecType = std::conditional_t<IsDecimalNumber<AttributeType>, ColumnDecimal<AttributeType>, ColumnVector<AttributeType>>;
+    using ColType = std::conditional_t<std::is_same<AttributeType, String>::value, ColumnString, ColVecType>;
+    const auto column = typeid_cast<const ColType *>(attributes[attribute_ind].get());
+    if (!column)
+        throw Exception{"An attribute should be a column of its type", ErrorCodes::BAD_ARGUMENTS};
+    for (const auto i : ext::range(0, points.size()))
+    {
+        size_t id = 0;
+        const auto found = find(points[i], id);
+        id = ids[id];
+        if (!found)
+        {
+            set_value(i, static_cast<OutputType>(get_default(i)));
+            continue;
+        }
+        if constexpr (std::is_same<AttributeType, String>::value)
+            set_value(i, static_cast<OutputType>(column->getDataAt(id)));
+        else
+            set_value(i, static_cast<OutputType>(column->getElement(id)));
+    }
+
+    query_count.fetch_add(points.size(), std::memory_order_relaxed);
+}
+
+namespace
+{
+
+inline void makeDifferences(IColumn::Offsets & values)
+{
+    for (size_t i = 1; i < values.size(); ++i)
+        values[i] -= values[i - 1];
+}
+
+struct Offset
+{
+    Offset() = default;
+
+    IColumn::Offsets ring_sizes;
+    IColumn::Offsets polygon_sizes;
+    IColumn::Offsets multi_polygon_sizes;
+
+    IColumn::Offset current_ring = 0;
+    IColumn::Offset current_polygon = 0;
+    IColumn::Offset current_multi_polygon = 0;
+};
+
+struct Data
+{
+    std::vector<IPolygonDictionary::Polygon> & dest;
+    std::vector<size_t> & ids;
+
+    void addPolygon(bool new_multi_polygon = false)
+    {
+        dest.emplace_back();
+        ids.push_back((ids.empty() ? 0 : ids.back() + new_multi_polygon));
+    }
+};
+
+void addNewMultiPolygon(Data &, Offset & offset)
+{
+    ++offset.current_multi_polygon;
+    if (offset.current_multi_polygon == offset.multi_polygon_sizes.size())
+        throw Exception{"Incorrect polygon formatting", ErrorCodes::BAD_ARGUMENTS};
```

Comment:
Also this error seem to be misleading. Imagine if user will get this error, what should they do? :)
I think it is more like assert. If so, you can add assert in debug mode. Or throw `LOGICAL_ERROR`.
(I just have no idea how user can specify incorrect array).

Reply:
Btw, what would happen if we will get empty arrays or single-point arrays?

Reply:
Yes, I think it is physically impossible to get this and some of the other error messages. Removed them. 
The only issues arising in a correctly-formatted array are with polygons containing less than 3 points. They can be stored as boost polygons, but it doesn't seem to handle them correctly in terms of our queries.
I think it makes more sense to just forbid any rings with less than 3 points.